backoffconnector.go

```diff
@@ -0,0 +1,87 @@
+package discovery
+
+import (
+	"context"
+	lru "github.com/hashicorp/golang-lru"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+type BackoffConnector struct {
+	cache      *lru.TwoQueueCache
+	host       host.Host
+	connTryDur time.Duration
+	backoff    BackoffFactory
+	mux        sync.Mutex
+}
+
+func NewBackoffConnector(h host.Host, cacheSize int, connectionTryDuration time.Duration, backoff BackoffFactory) (*BackoffConnector, error) {
```

Comment:
docstring; what is this?

Reply:
Will add docstrings. This is a utility that tries to connect to peers, but doesn't if we've already tried to connect to them recently (i.e. within the backoff period).

Reply:
docstring added

Reply:
plan is to use this for discovery at least within go-libp2p-pubsub